### PR TITLE
:warning: clusterctlv2 add low-level library for the provider components YAML

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR adds the low-level library for the provider components YAML

This is one of the core elements of clusterctl v2, and implements the management of the YAML defining components (CRD, Controller, RBAC) read from the provider repositories

It is important to notice that clusterctl applies a set of processing steps to the “raw” component read
from the provider repositories:
1. Checks for all the variables in the component YAML file and replace with corresponding config values
2. Ensures all the provider components are deployed in the target namespace (apply only to namespaced objects)
3. Ensures all the ClusterRoleBinding which are referencing namespaced objects have the name prefixed with the namespace name
4. Sets the watching namespace for the provider controller
5. Adds labels to all the components in order to allow easy identification of the provider objects

**Which issue(s) this PR fixes**:
Rif #1729 

/assign @vincepri 
/cc @detiber @ncdc

### CODE REVIEW

File: cmd/clusterctl/pkg/client/repository/components_client_test.go [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+					WithFile("v1.0.0", "components.yaml", util.JoinYaml(podYaml, configMapYaml)),
+				configVariablesClient: test.NewFakeVariableClient().WithVar(variableName, variableValue),
+			},
+			args: args{
+				version:           "v2.0.0",
+				targetNamespace:   "",
+				watchingNamespace: "",
+			},
+			wantErr: true,
+		},
```
Comment (Arvinderpal) [inline cmd/clusterctl/pkg/client/repository/components_client_test.go]: nit: consider adding test cases for `watchingNamespace` similar to `targetNamespace`

File: cmd/clusterctl/pkg/client/repository/components_test.go [hunk 2]
```diff
@@ -0,0 +1,566 @@
...
+			},
+			want: []string{"A", "B", "C", "D"},
+		},
+		{
+			name: "",
+			args: args{
+				data: "yaml with ${A} ${A} ${A}",
+			},
+			want: []string{"A"},
+		},
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/repository/components_test.go]: done

File: cmd/clusterctl/pkg/client/repository/components_client_test.go [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+				t.Errorf("Get().TargetNamespace() got = %v, want = %v ", got.TargetNamespace(), tt.want.targetNamespace)
+			}
+
+			if got.WatchingNamespace() != tt.want.watchingNamespace {
+				t.Errorf("Get().WatchingNamespace() got = %v, want = %v ", got.WatchingNamespace(), tt.want.watchingNamespace)
+			}
+
+			if !reflect.DeepEqual(got.Variables(), tt.want.variables) {
+				t.Errorf("Get().Variables() got = %v, want = %v ", got.WatchingNamespace(), tt.want.watchingNamespace)
+			}
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/repository/components_client_test.go]: done

File: cmd/clusterctl/pkg/client/repository/components_client_test.go (hunk 1 above)
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/repository/components_client_test.go]: done

**Comment by fabriziopandini** [comment]
@Arvinderpal thanks!
Comment addressed

File: cmd/clusterctl/pkg/client/repository/components.go [hunk 4]
```diff
@@ -64,3 +74,409 @@ type Components interface {
...
+func replaceVariables(yaml []byte, variables []string, configVariablesClient config.VariablesClient) ([]byte, error) {
+	tmp := string(yaml)
+	var missingVariables []string
+	for _, key := range variables {
+		val, err := configVariablesClient.Get(key)
+		if err != nil {
+			missingVariables = append(missingVariables, key)
+			continue
+		}
+		exp := regexp.MustCompile(`\$\{\s*` + key + `\s*\}`)
```
Comment (vincepri) [inline cmd/clusterctl/pkg/client/repository/components.go]: Maybe let's make this a different function so we can test it separately?

File: cmd/clusterctl/pkg/internal/util/cmd.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+	return lines, err
+}
+
+func (c *Cmd) Stdin(in io.Reader) *Cmd {
+	c.stdin = in
+	return c
+}
+
+func (c *Cmd) runInnerCommand() error {
+	cmd := exec.Command(c.command, c.args...) //nolint
```
Comment (vincepri) [inline cmd/clusterctl/pkg/internal/util/cmd.go]:
```suggestion
	cmd := exec.Command(c.command, c.args...) //nolint:gosec
```

File: cmd/clusterctl/pkg/internal/util/yaml.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+	}
+
+	r := bytes.Join(b, yamlSeparator)
+	r = bytes.TrimPrefix(r, cr)
+	r = bytes.TrimSuffix(r, cr)
+
+	return r
+}
+
+func ToUnstructured(rawyaml []byte) ([]unstructured.Unstructured, error) {
```
Comment (vincepri) [inline cmd/clusterctl/pkg/internal/util/yaml.go]: Add a comment

File: cmd/clusterctl/pkg/internal/util/yaml.go [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+	"bytes"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"sigs.k8s.io/yaml"
+)
+
+var yamlSeparator = []byte("---")
+
+func JoinYaml(yamls ...[]byte) []byte {
```
Comment (vincepri) [inline cmd/clusterctl/pkg/internal/util/yaml.go]: Add comment

File: cmd/clusterctl/pkg/client/repository/components.go (hunk 4 above)
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/repository/components.go]: This function is a simple a for loop around a replace and it is already tested. 
I don't think we will get a big improvement in splitting out the replace, but I will be happy to add more test if necessary

File: cmd/clusterctl/pkg/client/repository/components.go [hunk 8]
```diff
@@ -64,3 +74,409 @@ type Components interface {
...
+		Provider:          provider,
+		version:           version,
+		variables:         variables,
+		targetNamespace:   targetNamespace,
+		watchingNamespace: watchingNamespace,
+		objs:              objs,
+	}, nil
+}
+
+var variableRegEx = regexp.MustCompile(`\${\s*([A-Z0-9_]+)\s*}`)
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/repository/components.go]: do

File: cmd/clusterctl/pkg/client/repository/components.go [hunk 9]
```diff
@@ -64,3 +74,409 @@ type Components interface {
...
+	match := variableRegEx.FindAllStringSubmatch(string(data), -1)
+
+	for _, m := range match {
+		submatch := m[1]
+		if _, ok := variables[submatch]; !ok {
+			variables[submatch] = ""
+		}
+	}
+
+	var ret []string // nolint
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/repository/components.go]: done

File: cmd/clusterctl/pkg/client/repository/components.go [hunk 10]
```diff
@@ -64,3 +74,409 @@ type Components interface {
...
+		targetNamespace:   targetNamespace,
+		watchingNamespace: watchingNamespace,
+		objs:              objs,
+	}, nil
+}
+
+var variableRegEx = regexp.MustCompile(`\${\s*([A-Z0-9_]+)\s*}`)
+
+func inspectVariables(data []byte) []string {
+	variables := map[string]string{}
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/repository/components.go]: done

File: cmd/clusterctl/pkg/client/repository/components.go [hunk 11]
```diff
@@ -64,3 +74,409 @@ type Components interface {
...
+	match := variableRegEx.FindAllStringSubmatch(string(data), -1)
+
+	for _, m := range match {
+		submatch := m[1]
+		if _, ok := variables[submatch]; !ok {
+			variables[submatch] = ""
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/repository/components.go]: done

File: cmd/clusterctl/pkg/client/repository/components.go (hunk 4 above)
Comment (vincepri) [reply cmd/clusterctl/pkg/client/repository/components.go]: Yeah, I was more talking about the regex generation

**Pull request merged**